classes-recipe/image_types_qcom.bbclass

```diff
@@ -0,0 +1,95 @@
+
+inherit image_types
+
+IMAGE_TYPES += "qcomflash"
+
+QCOM_ESP_IMAGE ?= "esp-qcom-image"
+QCOM_ESP_FILE ?= "${@'efi.bin' if d.getVar('QCOM_ESP_IMAGE') else ''}"
+IMAGE_QCOMFLASH_ESPIMG ?= "${DEPLOY_DIR_IMAGE}/${QCOM_ESP_IMAGE}-${MACHINE}.rootfs.vfat"
+
+# There is currently no upstream-compatible way for the firmware to
+# identify and load the correct DTB from a combined-dtb that contains all
+# dtbs defined in KERNEL_DEVICETREE, so pick the first individual image
+# generated by linux-qcom-dtbbin, unless a different one is specified.
+QCOM_DTB_DEFAULT ?= "${@os.path.basename(d.getVar('KERNEL_DEVICETREE').split()[0][:-4]) if d.getVar('KERNEL_DEVICETREE') else ''}"
+QCOM_DTB_FILE ?= "dtb.bin"
+IMAGE_QCOMFLASH_DTB ?= "${DEPLOY_DIR_IMAGE}/dtb-${QCOM_DTB_DEFAULT}-image.vfat"
+
+QCOM_ROOTFS_FILE ?= "system.img"
+IMAGE_QCOMFLASH_FS_TYPE ??= "ext4"
+IMAGE_QCOMFLASH_ROOTFS ?= "${IMGDEPLOYDIR}/${IMAGE_LINK_NAME}.${IMAGE_QCOMFLASH_FS_TYPE}"
+
+QCOMFLASH_DIR = "${WORKDIR}/qcomflash"
+IMAGE_CMD:qcomflash = "create_qcomflash_pkg"
+do_image_qcomflash[dirs] = "${QCOMFLASH_DIR}"
+do_image_qcomflash[cleandirs] = "${QCOMFLASH_DIR}"
+do_image_qcomflash[depends] += "qcom-gen-partition-bins:do_deploy virtual/kernel:do_deploy \
+				${@'${QCOM_ESP_IMAGE}:do_image_complete' if d.getVar('QCOM_ESP_IMAGE') != '' else  ''}"
+IMAGE_TYPEDEP:qcomflash += "${IMAGE_QCOMFLASH_FS_TYPE}"
+
+create_qcomflash_pkg() {
+    local oldpwd="$PWD"
+    cd "${QCOMFLASH_DIR}"
+
+    # esp image
+    if [ -n "${QCOM_ESP_FILE}" ]; then
+        install -m 0644 ${IMAGE_QCOMFLASH_ESPIMG} ${QCOM_ESP_FILE}
+    fi
+
+    # dtb image
+    if [ -n "${QCOM_DTB_DEFAULT}" ] && [ -f "${IMAGE_QCOMFLASH_DTB}" ]; then
+        # default image
+        install -m 0644 ${IMAGE_QCOMFLASH_DTB} ${QCOM_DTB_FILE}
+        # copy all images so they can be made available via the same tarball
+        for dtbimg in ${DEPLOY_DIR_IMAGE}/dtb-*-image.vfat; do
+            install -m 0644 ${dtbimg} .
+        done
+    fi
+
+    # rootfs image
+    if [ -f "${IMAGE_QCOMFLASH_ROOTFS}" ]; then
+        install -m 0644 ${IMAGE_QCOMFLASH_ROOTFS} ${QCOM_ROOTFS_FILE}
+    fi
+
+    # partition bins
+    for gptm in ${DEPLOY_DIR_IMAGE}/gpt_main[0-9].bin; do
+        install -m 0644 ${gptm} .
+    done
+    for gpback in ${DEPLOY_DIR_IMAGE}/gpt_backup[0-9].bin; do
+        install -m 0644 ${gpback} .
+    done
+    for rawpg in ${DEPLOY_DIR_IMAGE}/rawprogram[0-9].xml; do
+        install -m 0644 ${rawpg} .
+    done
+    for patchfile in ${DEPLOY_DIR_IMAGE}/patch*.xml; do
+        install -m 0644 ${patchfile} .
+    done
+    if [ -f ${DEPLOY_DIR_IMAGE}/logfs_ufs_8mb.bin ]; then
+        install -m 0644 ${DEPLOY_DIR_IMAGE}/logfs_ufs_8mb.bin logfs_ufs_8mb.bin
+    fi
+    if [ -f ${DEPLOY_DIR_IMAGE}/zeros_5sectors.bin ]; then
+        install -m 0644 ${DEPLOY_DIR_IMAGE}/zeros_5sectors.bin zeros_5sectors.bin
+    fi
+
+    # boot firmware
+    for elffile in ${DEPLOY_DIR_IMAGE}/*.elf; do
```

Comment:
This breaks off there are no elf, mbn or fv files. Could you please implement something more future-proof? In my opinion two best options are using find or ls &&

Reply:
zeros_5sectors.bin is also hardcoded. the file has this name for UFS (well, 4K block size), but it would be zeros_33sectors.bin for disk with 512 block size.

Reply:
Updated to use for foo in `find`.